src/datahike/db.cljc

```diff
@@ -184,55 +220,36 @@
-                         (->> (-slice eavt (datom e a nil tx0) (datom e a nil txmax) :eavt) ;; e a _ tx
-                              (filter (fn [^Datom d] (= tx (datom-tx d)))))
-                         (-slice eavt (datom e a nil tx0) (datom e a nil txmax) :eavt) ;; e a _ _
-                         (->> (-slice eavt (datom e nil nil tx0) (datom e nil nil txmax) :eavt) ;; e _ v tx
-                              (filter (fn [^Datom d] (and (= v (.-v d))
-                                                          (= tx (datom-tx d))))))
-                         (->> (-slice eavt (datom e nil nil tx0) (datom e nil nil txmax) :eavt) ;; e _ v _
-                              (filter (fn [^Datom d] (= v (.-v d)))))
-                         (->> (-slice eavt (datom e nil nil tx0) (datom e nil nil txmax) :eavt) ;; e _ _ tx
-                              (filter (fn [^Datom d] (= tx (datom-tx d)))))
-                         (-slice eavt (datom e nil nil tx0) (datom e nil nil txmax) :eavt) ;; e _ _ _
-                         (if (indexing? db a) ;; _ a v tx
-                           (->> (-slice avet (datom e0 a v tx0) (datom emax a v txmax) :avet)
-                                (filter (fn [^Datom d] (= tx (datom-tx d)))))
-                           (->> (-slice aevt (datom e0 a nil tx0) (datom emax a nil txmax) :aevt)
-                                (filter (fn [^Datom d] (and (= v (.-v d))
-                                                            (= tx (datom-tx d)))))))
-                         (if (indexing? db a) ;; _ a v _
-                           (-slice avet (datom e0 a v tx0) (datom emax a v txmax) :avet)
-                           (->> (-slice aevt (datom e0 a nil tx0) (datom emax a nil txmax) :aevt)
-                                (filter (fn [^Datom d] (= v (.-v d))))))
-                         (->> (-slice aevt (datom e0 a nil tx0) (datom emax a nil txmax) :aevt) ;; _ a _ tx
-                              (filter (fn [^Datom d] (= tx (datom-tx d)))))
-                         (-slice aevt (datom e0 a nil tx0) (datom emax a nil txmax) :aevt) ;; _ a _ _
-                         (filter (fn [^Datom d] (and (= v (.-v d)) (= tx (datom-tx d)))) (-all eavt)) ;; _ _ v tx
-                         (filter (fn [^Datom d] (= v (.-v d))) (-all eavt)) ;; _ _ v _
-                         (filter (fn [^Datom d] (= tx (datom-tx d))) (-all eavt)) ;; _ _ _ tx
-                         (-all eavt)]))) ;; _ _ _ _
+           (let [[_ a _ _] pattern]
+             (search-indices eavt aevt avet pattern (indexing? db a))))
 
   IIndexAccess
   (-datoms [db index-type cs]
-           (let [index (get db index-type)]
-             (-slice index (components->pattern db index-type cs e0 tx0) (components->pattern db index-type cs emax txmax) index-type)))
+           (-slice (get db index-type)
+                   (components->pattern db index-type cs e0 tx0)
```

Comment:
Why the name `components->pattern`? If I understood well what it does is to `resolve-datom`.